### DISCUSSION TITLE
Intl.NumberFormat returning error, server and client side locales misaligned

### DISCUSSION THREAD

**Discussion by mbaric1** [vercel/next.js#19409] [Help] [answered]
# Bug report

## Describe the bug

I am trying to format number to locale string by setting explicit locale.
Intl.NumberFormat("hr-HR") returning following error: 
Warning: Text content did not match. Server: "€5,000.00" Client: "5.000,00 EUR"

Usage:
- [Number.prototype.toLocaleString()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toLocaleString)
- [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat)

## To Reproduce
Use following component: 

eur-to-hrk.tsx
``` ts
interface EurToHrkProps {
  price: number;
}

const EurToHrk = ({ price }: EurToHrkProps) => {
  return (
    <>
      {new Intl.NumberFormat("hr-HR", { style: "currency", currency: "EUR" }).format(price)}
      ~
      {new Intl.NumberFormat("hr-HR", { style: "currency", currency: "HRK" }).format(price * 7.52)}
    </>
  );
};

export default EurToHrk;
```

index.tsx
``` ts
const Home = () => {
  return (
    <>
      <Head>
        <title>Home</title>
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <EurToHrk price={5000}  />
    </>
  );
};
```

## Expected behavior

Server and client side to be aligned as explicit locale was passed.

## System information

- OS: Windows
- Browser: Chrome
- Version of Next.js: 9.5.5
- Version of Node.js: v12.16.2

**Comment by Paratron** [comment]
We had a similar error in a financial related product we currently develop. The problem is not with nextJS in this case, but with node itself.

The Intl number format is managed through [nodes internationalization modules](https://nodejs.org/api/intl.html). And - long story short - node runs by default only with the en-US settings when its comes to date/time formatting, money formatting and the likes.

You need to explicitly enable additional locales, or node will just default to the standard formatter.

There is [an npm package](https://www.npmjs.com/package/full-icu) which adds support for all ICU modules. After installing it, you need to tell the node process to use the ICU definitions from that package.

You can do that by adding an ENV variable to the start/dev/build scripts in the `package.json`:

```json
"scripts": {
        "test": "cross-env NODE_ICU_DATA=node_modules/full-icu jest",
        "dev": "cross-env NODE_ICU_DATA=node_modules/full-icu next",
        "build": "cross-env NODE_ICU_DATA=node_modules/full-icu next build",
        "start": "cross-env NODE_ICU_DATA=node_modules/full-icu next start",
    },
```
We also applied `cross-env` since our system might be executed on all kinds of mac/windows/linux computers of devs and we want it to work everywhere (setting envs in the command line is different everywhere).

Hope this fixes your problem!

**Comment by HaNdTriX** [comment, marked as answer]
the problem should be solved in node v14.x.x

**Reply by mbaric1** [reply]
Resolved by upgrading node to v14.15.1

**Comment by MonstraG** [comment]
Hit this problem with locale `nb-NO` and currency `NOK`:
`Warning: Text content did not match. Server: "100,00 NOK" Client: "kr 100,00"`
on node v16.2.0

Followed thread of issues, first in the [node github](https://github.com/nodejs/node/issues/39056), this is a [problem in V8](https://bugs.chromium.org/p/v8/issues/detail?id=11897)

**Comment by Klypalskyi** [comment]
I've also faced this issue with locale `uk` and currency `UAH`
```Warning: Text content did not match. Server: "160,00 ₴" Client: "160,00 грн"```

I'm using node 18.17.0

**Comment by zayaserg** [comment]
issue with locale uk-UA and currency UAH too
`Warning: Text content did not match. Server: "500 ₴" Client: "500 грн"`
node 21.6.1

**Comment by denvudd** [comment]
+1 with locale uk-UA and currency UAH
How can I fix that?

**Comment by SergeiKaptelin** [comment]
As a temporary solution, you can use [suppressHydrationWarning](https://react.dev/reference/react-dom/hydrate#suppressing-unavoidable-hydration-mismatch-errors) and looks like we should wait for fixes in future versions of node
